pyadjoint/checkpointing.py

```diff
@@ -0,0 +1,328 @@
+            )
+
+        if (
+            schedule.uses_storage_type(StorageType.DISK)
+            and not tape._package_data
+        ):
+            raise CheckpointError(
+                "The schedule employs disk checkpointing but it is not configured."
+            )
+        self.tape = tape
+        self._schedule = schedule
+        self.forward_schedule = []
+        self.reverse_schedule = []
+        self.timesteps = schedule.max_n
+        # This variable is used to indicate whether the adjoint model
+        self.adjoint_evaluated = False
+        self.mode = Mode.RECORD
+        self._current_action = next(self._schedule)
+        self.forward_schedule.append(self._current_action)
+        # Tell the tape to only checkpoint input data until told otherwise.
+        self.tape.latest_checkpoint = 0
+        self.end_timestep(-1)
+
+    def end_timestep(self, timestep):
+        """Mark the end of one timestep when taping the forward model.
+
+        Parameters
+        ----------
+        timestep : int
+            The current timestep.
+        """
+        if self.mode == Mode.EVALUATED:
+            raise CheckpointError("Not enough timesteps in schedule.")
+        elif self.mode != Mode.RECORD:
+            raise CheckpointError(f"Cannot end timestep in {self.mode}")
+        while not self.process_taping(self._current_action, timestep + 1):
+            self._current_action = next(self._schedule)
+            self.forward_schedule.append(self._current_action)
+
+    def end_taping(self):
+        """Process the end of the forward execution."""
+        current_timestep = self.tape.latest_timestep
+        while self.mode != Mode.EVALUATED:
+            self.end_timestep(current_timestep)
+            current_timestep += 1
+
+    @singledispatchmethod
+    def process_taping(self, cp_action, timestep):
+        """A single-dispatch generic function.
+        The process taping is used while the forward model is taped.
+
+        Parameters
+        ----------
+        cp_action : checkpoint_schedules.CheckpointAction
+            A checkpoint action from the schedule. The actions can
+            be `Forward`, `Reverse`, `EndForward`, `EndReverse`, `Copy`, or
+            `Move`.
+        timestep : int
+            The current timestep.
+
+        Returns
+        -------
+        bool
+            `True`, while a forward action is not finalised. Otherwise,
+            `False`.
+
+        Raises
+        ------
+        CheckpointError
+            If the checkpoint action is not supported.
+
+        Notes
+        -----
+        Additional details about checkpoint_schedules can be found at
+        checkpoint_schedules
+        `documentation <https://www.firedrakeproject.org/checkpoint_schedules/>`_.
+        """
+        raise CheckpointError(f"Unable to process {cp_action} while taping.")
+
+    @process_taping.register(Forward)
+    def _(self, cp_action, timestep):
```

Comment:
It would be nice to have some documentation of what the `process_taping` does when entering forward mode, or at least have a link to:
https://www.firedrakeproject.org/checkpoint_schedules/checkpoint_schedules.html#checkpoint_schedules.schedule.Forward